File: engine/src/main/java/com/terralite/engine/physics/Aabb.java
package com.terralite.engine.physics;

public record Aabb(double minX, double minY, double minZ, double maxX, double maxY, double maxZ) {
    public Aabb {
        if (maxX < minX) {
            throw new IllegalArgumentException("Max x cannot be less than min x");
        }
        if (maxY < minY) {
            throw new IllegalArgumentException("Max y cannot be less than min y");
        }
        if (maxZ < minZ) {
            throw new IllegalArgumentException("Max z cannot be less than min z");
        }
    }

    public boolean intersects(Aabb other) {
        return minX <= other.maxX && maxX >= other.minX
            && minY <= other.maxY && maxY >= other.minY
            && minZ <= other.maxZ && maxZ >= other.minZ;
    }
}

File: engine/src/commonMain/kotlin/com/terralite/engine/physics/Aabb.kt
package com.terralite.engine.physics

@JvmRecord
data class Aabb(
    val minX: Double, val minY: Double, val minZ: Double,
    val maxX: Double, val maxY: Double, val maxZ: Double
) {
    init {
        require(maxX >= minX) { "Max x cannot be less than min x" }
        require(maxY >= minY) { "Max y cannot be less than min y" }
        require(maxZ >= minZ) { "Max z cannot be less than min z" }
    }

    fun intersects(other: Aabb): Boolean =
        minX <= other.maxX && maxX >= other.minX &&
        minY <= other.maxY && maxY >= other.minY &&
        minZ <= other.maxZ && maxZ >= other.minZ
}
